src/frontend/js/components/ReferralDetailAssignment/AssignUnitModal.tsx

```diff
@@ -0,0 +1,224 @@
+import React, { useState, useEffect } from 'react';
+import { defineMessages, FormattedMessage } from 'react-intl';
+import ReactModal from 'react-modal';
+import { useUIDSeed } from 'react-uid';
+
+import { GenericErrorMessage } from 'components/GenericErrorMessage';
+import { Spinner } from 'components/Spinner';
+import { useReferralAction } from 'data';
+import { Referral, Unit } from 'types';
+
+const messages = defineMessages({
+  cancel: {
+    defaultMessage: 'Cancel',
+    description: 'Button to cancel assigning a unit and close the modal.',
+    id: 'components.ReferralDetail.AssignUnitModal.cancel',
+  },
+  formErrorMandatory: {
+    defaultMessage: 'An explanation is required when assigning a unit.',
+    description:
+      'Error message when the user assign a unit without providing an explanation.',
+    id: 'components.ReferralDetail.AssignUnitModal.formErrorMandatory',
+  },
+  formErrorServer: {
+    defaultMessage:
+      'There was an error while updating the referral. Please retry later or contact an administrator.',
+    description:
+      'Error message when the close referral in the modal fails for an unknown reason.',
+    id: 'components.ReferralDetail.AssignUnitModal.formErrorServer',
+  },
+  formLabelExplanation: {
+    defaultMessage: 'Assigning explanation',
+    description:
+      'Label for the assigning explanation field in the modal to close the referral.',
+    id: 'components.ReferralDetail.AssignUnitModal.formLabelExplanation',
+  },
+  modalTitle: {
+    defaultMessage: 'Assigning unit',
+    description: 'Title for the modal .',
+    id: 'components.ReferralDetail.AssignUnitModal.modalTitle',
+  },
+  modalMessage: {
+    defaultMessage:
+      'This action  will inform the unit <b>{ name }</b> of assigning this referral. You can provide an explanation to other stakeholders using the explanation field.',
+    description: 'Message warning user why close referral.',
+    id: 'components.ReferralDetail.AssignUnitModal.modalMessage',
+  },
+  update: {
+    defaultMessage: 'Assign unit',
+    description: 'Button to assign the unit in the relevant modal.',
+    id: 'components.ReferralDetail.AssignUnitModal.update',
+  },
+});
+
+// The `setAppElement` needs to happen in proper code but breaks our testing environment.
+// This workaround is not satisfactory but it allows us to both test <SendAnswerModal />
+// and avoid compromising accessibility in real-world use.
+const isTestEnv = typeof jest !== 'undefined';
+if (!isTestEnv) {
+  ReactModal.setAppElement('#app-root');
+}
+
+interface AssignUnitModalProps {
+  isAssignUnitModalOpen: boolean;
+  setIsAssignUnitModalOpen: (isOpen: boolean) => void;
+  referral: Referral;
+  unit: Unit;
+  isKeepDropdownMenu: boolean;
```

Comment:
This is not used in this component, it's not necessary to pass it as a prop. Just pass the setter (`setIsKeepDropdownMenu `)